docs/devel/luns.md

```diff
@@ -0,0 +1,56 @@
+Direct LUNs migration
+==============
+
+Summary
+-------
+
+VM contains disks. Some of them might be connected as a direct LUN.
+When a user wish to re-use the LUN disks, he needs them attached to the VM.
+The LUN should be used in a single place, or else, it will cause a data corruption.
+The storage where the LUNs reside, needs to be reachable to the VM environment.
+This feature page mostly about import from oVirt.
+
+Design
+------
+
+The requirements for the LUN migration code are (not in priority order)
+
+1. Gather the LUN informnation from the source destination.
+2. To ease on the user - less manual actions, more automated migration.
+3. Save unnecessary risks.
+
+Requirements
+----------------------------------------
+
+1. Destination (OCP cluster) reachability to the ISCSI storage, where the LUN is.
+2. RBAC permission to create `PersistentVolumes` by the controller.
+
+Risks and Mitigations
+----------------------------------------
+
+1. A LUN being used by a VM the user wish to migrate, without reachability to the destination.
+2. A VM that will be used while or after the migration with a direct LUN connected.
+   - A warning should be shown before starting the migration.
+   - (Optional) UI checkbox to the user, before starting the migration.
+3. An oVirt version with unsupported API. The [PR](https://github.com/oVirt/ovirt-engine/pull/559) provides LUN details is required.
+   - We may check for missing LUN details and skip the disk migration.
+   - (Optional) Have a mechanism to check the oVirt version for future work.
+4. Creating a `PersistentVolume` requires the file system type: `ext4` or `xfs` of the LUN, which we don't know.
+5. On non-local destination, the controller needs permission to create the `PersistentVolumes`.
+   - It may be enough to be added to the forklift-operator.
+
+Implementation
+----------------------------------------
+
+The controller gather data from source provider and saves them to the inventory.
+As part of the data gathering we will gather the LUN disks data. We have the VMs disk attachments where we follow to the disks.
+Within the disks API, if the _storage\_type_ is `LUN`, we will gather from the API endpoint `api/disks/<disk_id>` the size, address, IQN and the mapping of the LUN.
+Those will be found withing the `logical_unit` element of the `lun_storage`.
+On the migration plan a new steps will be added, creating a `PersistentVolume`(PV) and a `PersistentVolumeClaim`(PVC) per LUN disk.
+Both of them will contain the relevant data to use the LUN, the PVC will be bound the the PV by a unique label, `<vm_name>-vol-<volume-number>`.
+When the VM specification created, the controller will add a LUN disk using the PVC that was created as [described by kubevirt](https://kubevirt.io/user-guide/virtual_machines/disks_and_volumes/#lun).
```

Comment:
did you check if cdi supports defining luns? if it's not, wouldn't it make sense to add it there?

Reply:
all i could find is that CDI can create PVC. Which feels more effort as I need to put all the details there(as I need to the PVC). https://github.com/kubevirt/containerized-data-importer/blob/main/doc/datavolumes.md#target-storagepvc

Reply:
why would it require more efforts? wouldn't a data volume with 'source' that contains all the data you put into the PV be enough? Benny raised a valid point earlier today that if there's a cdi of the same storage type that we use in ovirt (e.g., netapp) it would make sense to use it here so that would be just another storage class that we specify on the pvc I suppose

Reply:
After discussing with Alexander Wels, CDI doesn't support it as I inspected. But he pointed out [this doc](https://github.com/kubevirt/kubevirt/blob/main/docs/iscsi-authentication.md) and his own words about the differences of my suggested approach and the doc above:

> Well that is slightly different, that takes a disk.img (or block volume) in a PVC and exposes it as a LUN inside the VM. It might not actually exist on an iSCSI device. It just looks like that to the VM. The example I pointed to is actually connecting to the iSCSI device and using the provided information to find the LUN on the device. You will notice that we need to provide an auth secret so the VM can connect to the device.
> Now it is possible to create a PV that points to an iSCSI device and then create a PVC that binds to the PV, and then use that PVC in the VM, and we can expose that PVC as a lun inside the VM, but that adds a bunch of extra layers vs just pointing directly to the LUN

Reply:
it's not that clear where that flow of exposing disk.img to a LUN inside the VM came from - that message is a bit out of context. but ok, that makes sense - let's try to do it that way. I also noticed that LUN is not part of the sources supported by CDI, but I was wondering if it would make sense to add it there rather than creating the PV and correlating it to the VM separately. if there's a way to do it in a k8s way, it's even better - let's check it

Reply:
CDI relies on `qemu-img` for converting the input VM image and only supports importing VM images to block disks or raw images. KubeVirt already supports direct LUNs but to my knowledge, we don't have a way to populate a direct LUN. We might be able to extend CDI to understand LUNs, but this will require some research

Reply:
Thanks Alice!